### PR TITLE
Fix ASB - Prevent the future from being stopped in production

### PR DESCRIPTION
**Pull request by da-kami**

### CODE REVIEW

**Comment by da-kami** [comment]
Note: I just did a successful swap between ASB (automated swap backedn - Alice) and CLI (Bob) on test/stagenet with this codebase :)

File: swap/src/bitcoin/wallet.rs [hunk 1]
```diff
@@ -163,14 +164,20 @@ impl BroadcastSignedTransaction for Wallet {
...
             tracing::debug!("found tx: {}", txid);
-            Ok(tx)
+
+            Result::<_, backoff::Error<Error>>::Ok(tx)
         })
         .await
-        .map_err(|err| anyhow!("transient errors to be retried: {:?}", err))
+        .map_err(|err| anyhow!("transient errors to be retried: {:?}", err))?;
+
+        Ok(tx)
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:180]: :+1: I also prefer this style.

File: swap/src/bitcoin/wallet.rs [hunk 2]
```diff
@@ -163,14 +164,20 @@ impl BroadcastSignedTransaction for Wallet {
...
+                .transaction_get(&txid)
+                .map_err(|_| backoff::Error::Transient(Error::NotYetMined))?;
             tracing::debug!("found tx: {}", txid);
-            Ok(tx)
+
+            Result::<_, backoff::Error<Error>>::Ok(tx)
         })
         .await
-        .map_err(|err| anyhow!("transient errors to be retried: {:?}", err))
+        .map_err(|err| anyhow!("transient errors to be retried: {:?}", err))?;
```
Comment (thomaseizinger) [inline swap/src/bitcoin/wallet.rs:178]: Can we `.context` here?

**Review by thomaseizinger — APPROVED** [review]

File: swap/src/bitcoin/wallet.rs [hunk 3]
```diff
@@ -163,14 +164,20 @@ impl BroadcastSignedTransaction for Wallet {
...
         tracing::debug!("watching for tx: {}", txid);
-        retry(ConstantBackoff::new(Duration::from_secs(1)), || async {
-            let client = Client::new(self.rpc_url.as_ref())?;
-            let tx = client.transaction_get(&txid)?;
+        let tx = retry(ConstantBackoff::new(Duration::from_secs(1)), || async {
+            let client = Client::new(self.rpc_url.as_ref())
+                .map_err(|err| backoff::Error::Permanent(Error::ElectrumConnection(err)))?;
+            let tx = client
+                .transaction_get(&txid)
+                .map_err(|_| backoff::Error::Transient(Error::NotYetMined))?;
```
[the remark this answers is not in the document]
Comment (da-kami) [reply swap/src/bitcoin/wallet.rs]: https://github.com/comit-network/xmr-btc-swap/pull/201/commits/2d660a3adde1461ba139baf1706e373808539cf2

I went through all the different errors and came to the conclusion that in this scenario only the `Protocol` error should be transient. However, I am not 100% sure about this.

File: swap/src/bitcoin/wallet.rs (hunk 2 above)
Comment (da-kami) [reply swap/src/bitcoin/wallet.rs]: done.

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**